File: lib/src/main/java/com/dhabensky/scopedvm/ScopedViewModelProvider.java
package com.dhabensky.scopedvm;

import com.dhabensky.scopedvm.subscription.OwnerSubscriptions;
import com.dhabensky.scopedvm.subscription.Subscription;
import com.dhabensky.scopedvm.subscription.Subscriptions;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelHack;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelStore;
import androidx.lifecycle.ViewModelStoreOwner;

/**
 * Created on 30.03.2019.
 * @author dhabensky <dhabensky@yandex.ru>
 */
public class ScopedViewModelProvider {

	private final ViewModelStoreOwner storeOwner;
	private final ViewModelProvider.Factory factory;
	private final String scope;
	private final Fragment requester;
	private ViewModelProvider provider;

	public ScopedViewModelProvider(@NonNull ViewModelStoreOwner storeOwner,
	                               @NonNull ViewModelProvider.Factory factory,
	                               @Nullable String scope,
	                               @NonNull Fragment requester) {
		this.storeOwner = storeOwner;
		this.factory = factory;
		this.scope = scope;
		this.requester = requester;
	}


	@MainThread
	public <T extends ViewModel> T get(@NonNull Class<T> modelClass) {
		return get(ViewModelHack.defaultKey(modelClass), modelClass);
	}

	@MainThread
	public <T extends ViewModel> T get(@NonNull String key, @NonNull Class<T> modelClass) {
		return getProvider().get(key, modelClass);
	}


	private ViewModelStore getScopedStore() {
		if (scope != null) {
			OwnerSubscriptions owner = Subscriptions.owner(storeOwner);
			// subscriptions
			Subscription subscription = new Subscription(owner, scope);
			owner.addSubscription(subscription);
			Subscriptions.user(requester).addSubscription(subscription);

			return owner.getScopedStore(scope);
		}
		else {
			return storeOwner.getViewModelStore();
		}
	}

	private ViewModelProvider getProvider() {
		if (provider == null) {
			provider = new ViewModelProvider(getScopedStore(), factory);
		}
		return provider;
	}

}

File: lib/src/main/java/com/dhabensky/svm/ScopedViewModelProvider.kt
package com.dhabensky.svm

import androidx.annotation.MainThread
import androidx.fragment.app.Fragment
import androidx.lifecycle.*
import com.dhabensky.svm.subscription.VMSubscription
import com.dhabensky.svm.subscription.VMSubscriptions

/**
 * Created on 16.03.2019.
 * @author dhabensky <dhabensky@yandex.ru>
 */
class ScopedViewModelProvider(
    private val storeOwner: ViewModelStoreOwner,
    private val factory: ViewModelProvider.Factory,
    private val scope: String?,
    private val requester: Fragment
) {

    @MainThread
    fun <T : ViewModel> get(modelClass: Class<T>): T {
        return get(defaultKey(modelClass), modelClass)
    }

    @MainThread
    fun <T : ViewModel> get(key: String, modelClass: Class<T>): T {
        return provider.get(key, modelClass)
    }

    private val provider by lazy {
        ViewModelProvider(getScopedStore(), factory)
    }

    private fun getScopedStore(): ViewModelStore {
        if (scope != null) {
            val owner = VMSubscriptions.owner(storeOwner)
            // subscriptions
            val subscription = VMSubscription(owner, scope)
            owner.addSubscription(subscription)
            VMSubscriptions.user(requester).addSubscription(subscription)

            return owner.getScopedStore(scope)
        }
        else {
            return storeOwner.viewModelStore
        }
    }

}
